src/System.Net.Primitives/src/System/Net/IPAddress.cs

```diff
@@ -609,12 +610,19 @@ public override int GetHashCode()
         {
             // For IPv6 addresses, we cannot simply return the integer
             // representation as the hashcode. Instead, we calculate
-            // the hashcode from the string representation of the address.
+            // the hashcode from the address literals and ScopeId.
             if (IsIPv6)
             {
                 if (_hashCode == 0)
                 {
-                    _hashCode = StringComparer.OrdinalIgnoreCase.GetHashCode(ToString());
+                    int hashCode = HashHelpers.Combine(HashHelpers.RandomSeed, unchecked((int)PrivateScopeId));
+                    ushort[] numbers = _numbers;
+                    for (int i = 0; i < numbers.Length; i += 2)
+                    {
+                        hashCode = HashHelpers.Combine(hashCode, numbers[i] << 16 | numbers[i + 1]);
+                    }
+
+                    _hashCode = hashCode;
```

Comment:
Do we have any tests today that cover IPAddress.GetHashCode? e.g. validating that equal addresses have the same hash code, validating IPAddresses created in various ways and with various settings still successfully generate a hash code, etc.?

Reply:
We absolutely need this before merge a change to GetHashCode.